### PR TITLE
Feature/55

### PR DESCRIPTION
**Pull request by StefanDywersant**
#55 User profile editor.

Feature/54 was merged to this feature-branch in order to use demo environment (www/).

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,8 +12,12 @@ RUN yarn
...
 
 # Bundle app source by overwriting all WORKDIR content.
-COPY . .
+COPY . tmp
 
+# Change owner since COPY/ADD assignes UID/GID 0 to all copied content.
+RUN chown -Rf node:node tmp; rsync -a tmp/ ./ && rm -rf tmp
+
+# Build client components
 RUN npm run build:client
```
Comment (gr4per) [inline Dockerfile]: not clear why we need this...

File: src/client/components/Alert/Alert.react.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import React, { Component, PropTypes } from 'react';
+import classNames from 'classnames';
+
+/**
+ * Draws basic alert message.
+ */
+export default class Alert extends Component {
+	static propTypes = {
```
Comment (gr4per) [inline src/client/components/Alert/Alert.react.js]: What is the difference to the existing approaches, why reinvent it ?

File: src/client/components/UserProfileEditor/UserProfileConstraints.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
+export default function(i18n) {
+	return {
```
Comment (gr4per) [inline src/client/components/UserProfileEditor/UserProfileConstraints.js]: It shows that we havnt really agreed on validation approach. But generally the idea is to use same rules on client and server and this here is a pure client side solution

File: src/client/components/UserProfileEditor/UserProfileEditor.react.js [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+	/**
+	 * Returns possible and translated salutations.
+	 * @returns {{value: string, label: string}[]}
+	 */
+	getSalutationOptions() {
+		return [
+			{ value: 'Mr', label: this.context.i18n.getMessage('UserProfileEditor.form.salutation.options.mr') },
+			{ value: 'Mrs', label: this.context.i18n.getMessage('UserProfileEditor.form.salutation.options.mrs') },
+			{ value: 'Ms', label: this.context.i18n.getMessage('UserProfileEditor.form.salutation.options.ms') }
```
Comment (gr4per) [inline src/client/components/UserProfileEditor/UserProfileEditor.react.js]: good idea but doesn't work in German. It is considered offensive. Me too

File: src/client/components/UserProfileEditor/UserProfileEditor.react.js [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+		return (
+			<div>
+				<h4 className="tab-description">{this.context.i18n.getMessage('UserProfileEditor.title', { userId : this.props.userId })}</h4>
+
+				<form className="form-horizontal" onSubmit={this.handleSubmit.bind(this)}>
+					{ this.renderField({
+						fieldName: 'languageId',
+						component: (
+							<LanguageField
```
Comment (gr4per) [inline src/client/components/UserProfileEditor/UserProfileEditor.react.js]: Difficult. This is the only field we have already supported via user menu directly.
Changing it requires updating the JWT (re-authenticate) and that is added complexity.
Two options: 
1. Leave it out of the profile editor
2. implement proper language switching (like via user menu)

File: Dockerfile (hunk 1 above)
Comment (StefanDywersant) [reply Dockerfile]: that build components according to webpack.production.js:
		profile: './src/client/components/UserProfileEditor/index.js',
		role: './src/client/components/UserRoleEditor/index.js'

so other services can access them via serviceRegistry

File: src/client/components/Alert/Alert.react.js (hunk 2 above)
Comment (StefanDywersant) [reply src/client/components/Alert/Alert.react.js]: My bad, I switched to service-base-ui's notification system and forgot to remove that component

File: src/client/components/UserProfileEditor/UserProfileEditor.react.js (hunk 5 above)
Comment (gr4per) [reply src/client/components/UserProfileEditor/UserProfileEditor.react.js]: Clarified: this is not only for self-editing (own profile) but also for editing other user's profile.
So we will cope with the fact that language change takes effect only after relogin when change in profile editor.

**Pull request merged**